persistent-test/src/PersistUniqueTest.hs

```diff
@@ -6,43 +6,139 @@ import Init
 
 -- mpsGeneric = False is due to a bug or at least lack of a feature in mkKeyTypeDec TH.hs
 share [mkPersist persistSettings { mpsGeneric = False }, mkMigrate "migration"] [persistLowerCase|
-  Fo
-      foo Int
-      bar Int
-      Primary foo
-      UniqueBar bar
-      deriving Eq Show
+
+Fo
+    foo Int
+    bar Int
+    Primary foo
+    UniqueBar bar
+    deriving Eq Show
+
+Ba
+    foo Int
+    baz Int
+    UniqueBaz baz
+    deriving Eq Show
+
+OnlyPrimaryKey
+    foo Int
+    name String
+    Primary foo
+    deriving Eq Show
+
 |]
 
+deriving stock instance Eq (Unique Fo)
+deriving stock instance Show (Unique Fo)
+
+deriving stock instance Show (Unique Ba)
+deriving stock instance Eq (Unique Ba)
+
+shouldCompile :: (OnlyOneUniqueKey OnlyPrimaryKey, AtLeastOneUniqueKey OnlyPrimaryKey) => IO ()
+shouldCompile = pure ()
+
 cleanDB :: (MonadIO m, PersistQuery backend, PersistEntityBackend Fo ~ backend) => ReaderT backend m ()
 cleanDB = do
   deleteWhere ([] :: [Filter Fo])
 
 specsWith :: Runner SqlBackend m => RunDb SqlBackend m -> Spec
-specsWith runDb = describe "custom primary key" $ do
-  it "getBy" $ runDb $ do
-    let b = 5
-    k <- insert $ Fo 3 b
-    Just vk <- get k
-    Just vu <- getBy (UniqueBar b)
-    vu @== Entity k vk
-  it "insertUniqueEntity" $ runDb $ do
-    let fo = Fo 3 5
-    Just (Entity _ insertedFoValue) <- insertUniqueEntity fo
-    Nothing <- insertUniqueEntity fo
-    fo @== insertedFoValue
-  it "checkUniqueUpdateable" $ runDb $ do
-    let f = 3
-    let b = 5
-    let fo = Fo f b
-    k <- insert fo
-    Just _ <- checkUnique fo -- conflicts with itself
-
-    let fo' = Fo (f + 1) b
-    Just _ <- checkUnique fo' -- conflicts with fo
-    Nothing <- checkUniqueUpdateable $ Entity k fo' -- but fo can be updated to fo'
-
-    let fo'' = Fo (f + 1) (b + 1)
-    insert_ fo''
-    Just (UniqueBar conflict) <- checkUniqueUpdateable $ Entity k fo'' -- fo can't be updated to fo''
-    conflict @== b + 1
```

Comment:
I think the old `insertUnique` code was actually broken on this, since it would be totally fine doing an `insert` based on the `UniqueBar` being actually unique, but then fail with the `Primary` key constraint violation.